sql/2021/security/cookie_max_age_expires_top_values.sql

```diff
@@ -0,0 +1,125 @@
+#standardSQL
+# Top 10 values of Max-Age and Expires cookie attributes.
+-- SQL Linter cannot handle DETERMINISTIC keyword so needs noqa ignore command on previous line
+CREATE TEMPORARY FUNCTION getCookieAgeValues(headers STRING, epochOfRequest NUMERIC)  -- noqa: PRS
+  RETURNS STRING DETERMINISTIC
+  LANGUAGE js AS '''
+  const regexMaxAge = new RegExp(/max-age\\s*=\\s*(?<value>-*[0-9]+)/i);
+  const regexExpires = new RegExp(/expires\\s*=\\s*(?<value>.*?)(;|$)/i);
+  const parsed_headers = JSON.parse(headers);
+  const cookies = parsed_headers.filter(h => h.name.match(/set-cookie/i));
+  const cookieValues = cookies.map(h => h.value);
+  const result = {
+      "maxAge": [],
+      "expires": []
+  };
+  cookieValues.forEach(cookie => {
+      let maxAge = null;
+      let expires = null;
+      if (regexMaxAge.exec(cookie)) {
+          maxAge = Number(regexMaxAge.exec(cookie)[1]);
+          result["maxAge"].push(maxAge);
+      }
+      if (regexExpires.exec(cookie)) {
+          expires = regexExpires.exec(cookie)[1];
+          result["expires"].push(expires);
+      }
+  });
+  return JSON.stringify(result);
+''';
+
+WITH age_values AS (
+  SELECT
+    client,
+    getCookieAgeValues(response_headers, startedDateTime) AS values
+  FROM
+    `httparchive.almanac.requests`
+  WHERE
+    date = "2021-07-01"
+),
+
+max_age_values AS (
+SELECT
+  client,
+  SUM(COUNT(0)) OVER (PARTITION BY client) AS total_max_age_values,
+  COUNT(0) AS max_age_count,
+  max_age_value
+FROM age_values,
+UNNEST(JSON_QUERY_ARRAY(values, "$.maxAge")) AS max_age_value
```

Comment:
You typically can't use the `SUM(COUNT(0)) OVER (PARTITION BY client)` construct when using `UNNEST` as values are double counted.

This is something we only discovered this year, so even if this query is copied from last year that still doesn't make it right.

Can you try rewriting this query with a separate total and see if you get the same numbers? I would guess not... If not then all the Partition queries might need to be rewritten.

(this comment applies throughout)

Reply:
Added some more info here: https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Analysts'-Guide#a-warning-on-totals

Reply:
I've rewritten four cases where `PARTITION BY` was used in combination with `UNNEST`:
- `csp_allowed_host_frequency.sql`: large impact on `total_pages` column (+/- 2.000 -> 10.000 for `total_pages` in `sample_data`). However, the 2020 article never stated what the mentioned percentage was relative to, so the error was never explicitly published.
- `sri_hash_functions.sql`: very small impact on `total_sri_elements` column (< 0.02 on `pct` in `sample_data`). These numbers were used in the 2020 article.
- `cookie_max_age_expires_top_values.sql`: No impact
- `sri_popular_hosts.sql`: No impact